// @lovable.dev/vite-tanstack-config/dist/index.js
// src/index.ts
import { componentTagger } from "lovable-tagger";
import { mergeConfig, loadEnv } from "vite";
var SANDBOX_ENV_VAR = "DEV_SERVER__PROJECT_PATH";
var LOVABLE_SANDBOX_ENV_VAR = "LOVABLE_SANDBOX";
function isSandboxEnvironment() {
  return process.env[LOVABLE_SANDBOX_ENV_VAR] === "1" || !!process.env[SANDBOX_ENV_VAR];
}
function logWarning(message) {
  console.warn(`[@lovable.dev/vite-tanstack-config] ${message}`);
}
function validateHeaders(config) {
  const headers = config.server?.headers;
  if (!headers)
    return;
  const headerKeys = Object.keys(headers);
  if (headerKeys.length > 0) {
    logWarning(`Your config includes server headers that will be removed in sandbox mode: ${headerKeys.join(", ")}`);
  }
}
function isExternalUrl(url) {
  try {
    const parsed = new URL(url);
    return !["localhost", "127.0.0.1", "::1", "0.0.0.0", "::"].includes(parsed.hostname);
  } catch {
    return false;
  }
}
function validateProxy(config) {
  const proxy = config.server?.proxy;
  if (!proxy)
    return;
  for (const [path, proxyConfig] of Object.entries(proxy)) {
    if (typeof proxyConfig === "string") {
      if (isExternalUrl(proxyConfig)) {
        logWarning(
          `Proxy '${path}' targets external URL: ${proxyConfig}. This may cause issues in the sandbox environment.`
        );
      }
    } else if (proxyConfig) {
      const target = typeof proxyConfig.target === "string" ? proxyConfig.target : void 0;
      if (target && isExternalUrl(target)) {
        logWarning(
          `Proxy '${path}' targets external URL: ${target}. This may cause issues in the sandbox environment.`
        );
      }
      if (proxyConfig.ws) {
        logWarning(`Proxy '${path}' enables WebSocket proxying, which may interfere with HMR.`);
      }
      if (proxyConfig.changeOrigin) {
        logWarning(`Proxy '${path}' changes origin, which may cause CORS issues.`);
      }
    }
  }
}
function validateCors(config) {
  if (config.server?.cors && typeof config.server.cors !== "boolean" && config.server.cors !== null) {
    logWarning("Custom CORS config detected. Sandbox requires specific CORS settings for proxy compatibility.");
  }
}
function validatePort(config) {
  if (config.server?.port && config.server.port !== 8080) {
    logWarning(`Port ${config.server.port} configured, but sandbox requires port 8080. Using 8080.`);
  }
}
function validateConfig(config) {
  validateHeaders(config);
  validateProxy(config);
  validateCors(config);
  validatePort(config);
}
function cleanServerConfig(config) {
  if (config.server?.headers || config.server?.cors || config.server?.proxy) {
    const { headers: _headers, cors: _cors, proxy: _proxy, ...serverRest } = config.server;
    return {
      ...config,
      server: serverRest
    };
  }
  return config;
}
function applyWatchDebounceDefaults(config) {
  const existingWatch = config.server?.watch ?? {};
  const existingAwaitWriteFinish = existingWatch.awaitWriteFinish;
  const hasAwaitWriteFinishObject = !!existingAwaitWriteFinish && typeof existingAwaitWriteFinish === "object" && !Array.isArray(existingAwaitWriteFinish);
  return mergeConfig(config, {
    server: {
      watch: {
        ...existingWatch,
        awaitWriteFinish: {
          ...hasAwaitWriteFinishObject ? existingAwaitWriteFinish : {},
          stabilityThreshold: 1e3,
          pollInterval: 100
        }
      }
    }
  });
}
var SSR_CAPTURE_KEY = "__LOVABLE_TANSTACK_CAPTURE_SSR_ERROR__";
function devSsrErrorLogger() {
  let lastCapture;
  const CAPTURE_TTL_MS = 5e3;
  const capture = (error) => {
    lastCapture = { error, at: Date.now() };
  };
  const consumeCapture = () => {
    if (!lastCapture)
      return void 0;
    if (Date.now() - lastCapture.at > CAPTURE_TTL_MS) {
      lastCapture = void 0;
      return void 0;
    }
    const { error } = lastCapture;
    lastCapture = void 0;
    return error;
  };
  return {
    name: "dev-ssr-error-logger",
    apply: "serve",
    configureServer(server) {
      globalThis[SSR_CAPTURE_KEY] = capture;
      const g = globalThis;
      if (typeof g.addEventListener === "function") {
        const addListener = g.addEventListener;
        addListener("error", (e) => capture(e.error ?? e));
        addListener("unhandledrejection", (e) => capture(e.reason));
      }
      const onUnhandledRejection = (reason) => capture(reason);
      process.on("unhandledRejection", onUnhandledRejection);
      server.httpServer?.once("close", () => {
        process.off("unhandledRejection", onUnhandledRejection);
      });
      server.middlewares.use((_req, res, next) => {
        const origEnd = res.end.bind(res);
        res.end = (...args) => {
          if (res.statusCode >= 500) {
            const captured = consumeCapture();
            let err;
            if (captured instanceof Error) {
              err = captured;
            } else if (typeof captured === "string" && captured.length > 0) {
              err = new Error(captured);
            } else {
              err = null;
            }
            try {
              server.ws.send({
                type: "custom",
                event: "server-ssr-error",
                data: err ? { name: err.name, message: err.message, stack: err.stack } : { name: "Error", message: "SSR rendering failed" }
              });
            } catch {
            }
          }
          return origEnd(...args);
        };
        next();
      });
    },
    transform(code, id) {
      const normalizedId = id.replace(/\\/g, "/");
      const isTargetModule = normalizedId.includes("/@tanstack/start-server-core/src/request-response.ts") || normalizedId.includes("/@tanstack/start-server-core/dist/esm/request-response.js");
      if (!isTargetModule) {
        return null;
      }
      const needle = "handler(request, requestOpts)";
      if (!code.includes(needle)) {
        return null;
      }
      return code.replace(
        needle,
        `Promise.resolve(${needle}).catch((err) => { globalThis.${SSR_CAPTURE_KEY}?.(err); throw err; })`
      );
    }
  };
}
function devServerFnErrorLogger() {
  const HMR_SEND_KEY = "__TANSTACK_SERVER_FN_HMR_SEND__";
  return {
    name: "dev-server-fn-error-logger",
    apply: "serve",
    enforce: "pre",
    configureServer(server) {
      globalThis[HMR_SEND_KEY] = (data) => {
        server.ws.send({
          type: "custom",
          event: "server-fn-error",
          data
        });
      };
    },
    transform(code, id) {
      const normalizedId = id.replace(/\\/g, "/");
      const isTargetModule = normalizedId.includes("/@tanstack/start-server-core/src/server-functions-handler.ts") || normalizedId.includes("/@tanstack/start-server-core/dist/esm/server-functions-handler.js");
      if (!isTargetModule) {
        return null;
      }
      const needle = "const unwrapped = res.result || res.error";
      if (!code.includes(needle)) {
        return null;
      }
      return code.replace(
        needle,
        `${needle}

      if (res?.error) {
        const err = res.error
        const payload = {
          source: 'tanstack',
          type: 'server-fn-error',
          method: request.method,
          url: request.url,
          name: err?.name ?? 'Error',
          message: err?.message ?? String(err),
          stack: typeof err?.stack === 'string' ? err.stack : undefined,
        }
        globalThis.${HMR_SEND_KEY}?.(payload)
      }`
      );
    }
  };
}
function defineConfig(configOrOptions = {}) {
  return async (env) => {
    const { command, mode } = env;
    const isSandbox = isSandboxEnvironment();
    let options;
    if (typeof configOrOptions === "function") {
      const userConfig = await Promise.resolve(configOrOptions(env));
      options = { vite: userConfig };
    } else if (configOrOptions instanceof Promise) {
      options = { vite: await configOrOptions };
    } else {
      const optionObject = configOrOptions && typeof configOrOptions === "object" ? configOrOptions : {};
      if ("cloudflare" in optionObject) {
        if (optionObject.cloudflare === false && !("nitro" in optionObject)) {
          optionObject.nitro = false;
          logWarning(
            "`cloudflare: false` is deprecated in v2 \u2014 migrated to `nitro: false`. Update your config to drop the warning."
          );
        } else {
          logWarning(
            "The `cloudflare` option was removed in v2. Use `nitro: false` to disable the deploy plugin, or `nitro: { preset: '...' }` to configure it. The `cloudflare` setting is being ignored."
          );
        }
      }
      const hasLovableKey = "vite" in optionObject || "serverFnErrorLogger" in optionObject || "ssrErrorLogger" in optionObject || "nitro" in optionObject || "tanstackStart" in optionObject || "react" in optionObject || "envDefine" in optionObject || "hmrGate" in optionObject;
      options = hasLovableKey ? optionObject : { vite: optionObject };
    }
    const internalPlugins = [];
    const tailwindcss = (await import("@tailwindcss/vite")).default;
    internalPlugins.push(tailwindcss());
    const tsConfigPaths = (await import("vite-tsconfig-paths")).default;
    internalPlugins.push(tsConfigPaths({ projects: ["./tsconfig.json"] }));
    if (options.serverFnErrorLogger !== false) {
      internalPlugins.push(devServerFnErrorLogger());
    }
    if (options.ssrErrorLogger !== false) {
      internalPlugins.push(devSsrErrorLogger());
    }
    const { tanstackStart } = await import("@tanstack/react-start/plugin/vite");
    const tanstackStartDefaults = {
      importProtection: {
        behavior: "error",
        client: {
          files: ["**/server/**"],
          specifiers: ["server-only"]
        }
      }
    };
    const tanstackStartOptions = mergeConfig(tanstackStartDefaults, options.tanstackStart ?? {});
    internalPlugins.push(tanstackStart(tanstackStartOptions));
    const explicitNitro = options.nitro === true || typeof options.nitro === "object" && options.nitro !== null;
    const shouldRunNitro = options.nitro !== false && command === "build" && (explicitNitro || isSandbox);
    if (options.nitro === void 0 && command === "build" && !isSandbox) {
      console.info(
        "[@lovable.dev/vite-tanstack-config] No Lovable context detected \u2014 skipping nitro deploy plugin. Pass `nitro: true` (or `nitro: { ... }`) to force-enable."
      );
    }
    if (shouldRunNitro) {
      let nitroMod;
      try {
        nitroMod = await import("nitro/vite");
      } catch (err) {
        if (explicitNitro || isSandbox) {
          const reason = explicitNitro ? `an explicit \`nitro\` option (${options.nitro === true ? "true" : "{ ... }"}) was set in vite.config.ts` : "running inside a Lovable sandbox";
          throw new Error(
            `[@lovable.dev/vite-tanstack-config] ${reason}, but the \`nitro\` package isn't installed. Add \`nitro\` to your devDependencies (matches peerDependency >=3.0.260429-beta). Original import error: ${err instanceof Error ? err.message : String(err)}`
          );
        }
      }
      if (nitroMod) {
        const { nitro } = nitroMod;
        const userNitroOpts = typeof options.nitro === "object" && options.nitro ? options.nitro : {};
        const preset = userNitroOpts.preset ?? process.env.NITRO_PRESET ?? "cloudflare-module";
        const output = {
          dir: "dist",
          serverDir: "dist/server",
          publicDir: "dist/client",
          ...userNitroOpts.output
        };
        const nitroOpts = { ...userNitroOpts, preset, output };
        if (preset === "cloudflare-module") {
          nitroOpts.cloudflare = { nodeCompat: true, deployConfig: true, ...userNitroOpts.cloudflare };
        }
        internalPlugins.push(nitro(nitroOpts));
      }
    }
    const viteReact = (await import("@vitejs/plugin-react")).default;
    internalPlugins.push(viteReact(options.react));
    if (command === "serve") {
      const hmrGateOpt = options.hmrGate;
      const enableHmrGate = isSandbox ? hmrGateOpt !== false : !!hmrGateOpt;
      if (enableHmrGate) {
        const { hmrGatePlugin } = await import("@lovable.dev/vite-plugin-hmr-gate");
        const hmrGateOptions = typeof hmrGateOpt === "object" ? hmrGateOpt : {};
        internalPlugins.push(hmrGatePlugin(hmrGateOptions));
      }
    }
    if (command === "serve" && isSandbox) {
      const { devServerBridgePlugin } = await import("@lovable.dev/vite-plugin-dev-server-bridge");
      internalPlugins.push(devServerBridgePlugin());
    }
    if (command === "serve" && mode === "development") {
      internalPlugins.push(componentTagger());
    }
    let envDefine = {};
    if (options.envDefine !== false) {
      const loadedEnv = loadEnv(mode, process.cwd(), "VITE_");
      for (const [key, value] of Object.entries(loadedEnv)) {
        envDefine[`import.meta.env.${key}`] = JSON.stringify(value);
      }
    }
    let config = {
      define: envDefine,
      // Match the build's CSS pipeline in dev. Vite uses PostCSS in dev and only
      // runs Lightning CSS at build, so build-time transforms (e.g. collapsing a
      // hand-written `-webkit-backdrop-filter` to the prefixed form Chrome ignores)
      // break the built/static output while the dev preview looks fine. Running
      // Lightning CSS in both keeps the preview honest.
      css: { transformer: "lightningcss" },
      resolve: {
        alias: {
          "@": `${process.cwd()}/src`
        },
        dedupe: [
          "react",
          "react-dom",
          "react/jsx-runtime",
          "react/jsx-dev-runtime",
          "@tanstack/react-query",
          "@tanstack/query-core"
        ]
      },
      plugins: [...internalPlugins, ...options.plugins ?? []]
    };
    if (options.vite) {
      config = mergeConfig(config, options.vite);
    }
    if (isSandbox) {
      validateConfig(config);
      config = cleanServerConfig(
        mergeConfig(config, {
          server: { host: "::", port: 8080, strictPort: true }
        })
      );
    } else {
      config = mergeConfig({ server: { host: "::", port: 8080 } }, config);
    }
    if (isSandbox) {
      return config;
    }
    return applyWatchDebounceDefaults(config);
  };
}
export {
  defineConfig
};
